### ISSUE TITLE
Analyzer throws unexpected error for typedef with unused type parameter in legacy code .

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#46178]
Dart SDK version: 2.14.0-152.0.dev (dev) (Tue May 25 15:03:31 2021 -0700) on "windows_x64"

The following source code example passes with dart and throws a compile time error with analyzer:
``` dart
// @dart = 2.9

typedef F<X> = void Function<Y extends X>();
F<X> toF<X>(X x) => throw "Should not reach here";

typedef G<X extends num> = void Function();

test(G source) {
  var fsource = toF(source);
  F<G<dynamic>> target1 = fsource;
}

main() {}
```
Please not that this is a legacy code, if I remove `// @dart = 2.9` comment line both dart and analyzer pass here.

Sample output is:
```
$> dart test.dart

$> dartanalyzer test.dart
Analyzing test.dart...
  error - 'dynamic' doesn't conform to the bound 'num' of the type parameter 'X'. - test.dart:10:7 - type_argument_not_matching_bounds
  hint - The value of the local variable 'target1' isn't used. - test.dart:10:17 - unused_local_variable
1 error and 1 hint found.
```

**Comment by eernstg** [comment]
I agree that the analyzer should not report a compile-time error stating that `G<dynamic>` has a bounds violation: The type parameter of `G` has no variance, so when we check whether it's correctly super-bounded we replace `dynamic` by `Null`, and confirm that `G<Null>` is regular-bounded, in `@dart = 2.9`.

There is no error in line 10 based on assignability, because `G<T>` just means `void Function()` no matter which type `T` we're passing as the type argument, so we _can_ initialize a variable of type `F<G<T1>>` with an expression of type `F<G<T2>>` for _any_ `T1` and `T2`, as long as none of those types is a compile-time error in itself.
